source/simulator/helper_functions.cc

```diff
@@ -1746,114 +1766,79 @@ namespace aspect
                 }
             }
 
           cell->get_dof_indices (local_dof_indices);
+          const unsigned int component_idx_T = introspection.component_indices.temperature;
 
-          // copy reaction rates and new values for the compositional fields
-          for (unsigned int j=0; j<dof_handler.get_fe().base_element(introspection.base_elements.compositional_fields).dofs_per_cell; ++j)
-            for (unsigned int c=0; c<introspection.n_compositional_fields; ++c)
-              {
-                const unsigned int composition_idx
-                  = dof_handler.get_fe().component_to_system_index(introspection.component_indices.compositional_fields[c],
-                                                                   /*dof index within component=*/ j);
-
-                // skip entries that are not locally owned:
-                if (dof_handler.locally_owned_dofs().is_element(local_dof_indices[composition_idx]))
-                  {
-                    distributed_vector(local_dof_indices[composition_idx]) = in_C.composition[j][c];
-                    distributed_reaction_vector(local_dof_indices[composition_idx]) = accumulated_reactions_C[j][c];
-                  }
-              }
-
-          // copy reaction rates and new values for the temperature field
-          for (unsigned int j=0; j<dof_handler.get_fe().base_element(introspection.base_elements.temperature).dofs_per_cell; ++j)
+          for (unsigned int dof_idx = 0; dof_idx < local_dof_indices.size(); ++dof_idx)
             {
-              const unsigned int temperature_idx
-                = dof_handler.get_fe().component_to_system_index(introspection.component_indices.temperature,
-                                                                 /*dof index within component=*/ j);
-
-              // skip entries that are not locally owned:
-              if (dof_handler.locally_owned_dofs().is_element(local_dof_indices[temperature_idx]))
+              const auto comp_pair = dof_handler.get_fe().system_to_component_index(dof_idx);
+              const unsigned int component_idx = comp_pair.first;
+              if (component_idx>=component_idx_T) // ignore velocity, pressure, etc.
                 {
-                  if (temperature_and_composition_use_same_fe)
-                    distributed_vector(local_dof_indices[temperature_idx]) = in_C.temperature[j];
-                  else
-                    distributed_vector(local_dof_indices[temperature_idx]) = in_T.temperature[j];
+                  // the how-manyth dof is that for the current component and where do we find this in the list
+                  // of evaluation points?
+                  const unsigned int index_within = comp_pair.second;
+                  const unsigned int field_index = component_idx-component_idx_T;
+                  const unsigned int idx = support_point_index_by_field[field_index][index_within];
```

Comment:
I find this hard to follow. I understand the general concept, but if you could add another sentence of explanation (either here, or where you fill the support_point_index_by_field vector) I think that would be helpful.

Reply:
better?